### PR TITLE
Pipeline

### PR DESCRIPTION
**Pull request by tsarn**
Pipeline for highlighting code snippets

### CODE REVIEW

File: highlighter/highlighter.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+func MakeHighlighter(storage types.SnippetStorage) Highlighter {
+	return Highlighter{
+		make(chan types.Snippet),
+		storage,
+	}
+}
+
+func highlightSnippet(snippet *types.Snippet) (string, error) {
+	out, err := exec.Command("pygmentize", "-l", string(snippet.Language), "-f", "html").Output()
```
Comment (M0nsteR) [inline highlighter/highlighter.go:22]: Could you detect any potential problems with this code?

Comment (M0nsteR) [inline highlighter/highlighter.go]: Is smth missing here? What will "pygmentize" pygmentize? This code looks suspicious in many ways. Could you please identify some potential risks here?

File: highlighter/highlighter.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	if err != nil {
+		return "", err
+	}
+	return string(out), nil
+}
+
+func (h *Highlighter) Run() {
+	for {
+		snippet := <-h.requests
```
Comment (M0nsteR) [inline highlighter/highlighter.go:31]: So, what we have is a single worker running jobs sequentially? Was that the intent of your pipeline?

File: main.go [hunk 3]
```diff
@@ -31,10 +32,15 @@ func main() {
 		panic(err)
 	}
 
+	h := highlighter.MakeHighlighter(postgres)
+
+	go h.Run()
```
Comment (M0nsteR) [inline main.go]: So, what we have here is a single worker, right?

**Review by M0nsteR — CHANGES_REQUESTED** [review]
You shouldn't mix refactoring code and pipeline implementation in the same PR.

File: highlighter/highlighter.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	"log"
+	"os/exec"
+)
+
+type Highlighter struct {
+	requests chan types.Snippet
+	storage  types.SnippetStorage
+}
+
+func MakeHighlighter(storage types.SnippetStorage) Highlighter {
```
Comment (NekoNe) [inline highlighter/highlighter.go]: Just `New`, please

File: highlighter/highlighter.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
+package highlighter
+
+import (
+	"github.com/mp-hl-2021/splinter/types"
+	"log"
+	"os/exec"
+)
+
+type Highlighter struct {
```
Comment (NekoNe) [inline highlighter/highlighter.go:14]: Do I understand it correctly?
You are trying to use an unbuffered channel as a message queue? Please, don't do that.

File: highlighter/highlighter.go [hunk 6]
```diff
@@ -11,15 +13,27 @@ type Highlighter struct {
 	storage  types.SnippetStorage
 }
 
-func MakeHighlighter(storage types.SnippetStorage) Highlighter {
+func New(storage types.SnippetStorage) Highlighter {
 	return Highlighter{
-		make(chan types.Snippet),
+		make(chan types.Snippet, 256),
```
Comment (NekoNe) [inline highlighter/highlighter.go]: Why 256? Why hardcoded? =)

Comment (tsarn) [reply highlighter/highlighter.go]: Seemed like a round number. Fixed.

File: usecases/delegated.go [hunk 7]
```diff
@@ -84,6 +86,8 @@ func (u DelegatedUserInterface) PostSnippet(author types.UserId, contents string
 		return types.Snippet{}, err
 	}
 
+	u.Highlighter.Post(snippet)
```
Comment (NekoNe) [inline usecases/delegated.go]: What will happen when the buffer in the buffered channel be full?

Comment (tsarn) [reply usecases/delegated.go]: Added a warning, don't really know what to do besides that.

Comment (NekoNe) [reply usecases/delegated.go]: It is actually a design fault. You need an actual queue between API call and Highlighter. The easiest way to do it is to store an event into DB at the API call and reading events from DB with some goroutine (actual pipeline start).

**Review by NekoNe — APPROVED** [review]
Please, pay attention to this https://github.com/mp-hl-2021/splinter/blob/0d15b74f16117e5194d9571badcd384cdfeb7375/highlighter/highlighter.go#L46

The idea behind the pipeline was to reduce coupling between CPU-bound and IO-bound jobs. In your case with `exec python`, it is slightly different, but the idea is generally the same.

Ok, let's ship it.

**Pull request merged**